datastore/zmq/enabled.go

```diff
@@ -0,0 +1,188 @@
+//go:build has_zmq
+
+package zmq
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pebbe/zmq4"
+	"github.com/sirupsen/logrus"
+	"github.com/teslamotors/fleet-telemetry/metrics"
+	"github.com/teslamotors/fleet-telemetry/metrics/adapter"
+	"github.com/teslamotors/fleet-telemetry/telemetry"
+)
+
+// Metrics stores metrics reported from this package
+type Metrics struct {
+	errorCount   adapter.Counter
+	publishCount adapter.Counter
+	byteTotal    adapter.Counter
+}
+
+var (
+	metricsRegistry Metrics
+	metricsOnce     sync.Once
+)
+
+// ZMQProducer implements the telemetry.Producer interface by publishing to a
+// bound zmq socket.
+type ZMQProducer struct {
+	topicPrefix string
+	topicSuffix string
+	topicSep    string
+	ctx         context.Context
+	sock        *zmq4.Socket
+	logger      *logrus.Logger
+}
+
+// Publish the record to the socket.
+func (p *ZMQProducer) Produce(rec *telemetry.Record) {
+	if p.ctx.Err() != nil {
+		return
+	}
+
+	topicParts := make([]string, 0, 4)
+	if len(p.topicPrefix) > 0 {
+		topicParts = append(topicParts, p.topicPrefix)
+	}
+	topicParts = append(topicParts, rec.Vin)
+	topicParts = append(topicParts, rec.TxType)
+	if len(p.topicSuffix) > 0 {
+		topicParts = append(topicParts, p.topicSuffix)
+	}
+	joinChar := DefaultTopicSep
+	if len(p.topicSep) > 0 {
+		joinChar = p.topicSep
+	}
+	topic := strings.Join(topicParts, joinChar)
+
+	if nBytes, err := p.sock.SendMessage(topic, rec.Payload()); err != nil {
+		metricsRegistry.errorCount.Inc(map[string]string{"record_type": rec.TxType})
+		p.logger.Errorf("Failed sending log on zmq socket: %s", err.Error())
+	} else {
+		metricsRegistry.byteTotal.Add(int64(nBytes), map[string]string{"record_type": rec.TxType})
+		metricsRegistry.publishCount.Inc(map[string]string{"record_type": rec.TxType})
+	}
+}
+
+// NewProducer creates a ZMQProducer with the given config.
+func NewProducer(ctx context.Context, config *Config, metrics metrics.MetricCollector, logger *logrus.Logger) (producer telemetry.Producer, err error) {
```

Comment:
can we add an integration test which verifies the producer logic is working as expected? There are some integration test done already for kafka, kinesis and pubsub

Reply:
Will do. FWIW zmq has long build times, if you're OK with increasing the time it takes to run the tests

Reply:
Can you expand more on what errors you are getting on `make generate-certs`?
Also cant speak much about using local IPs for integration test. Can you try with adding a zmq service in docker-compose.yml and use that service instead?

Reply:
```
jordan@jordan:~/fleet-telemetry$ make generate-certs
go run tools/main.go
2023/10/06 10:24:31 Creating Root Signing Cert for: Tesla Motors Products CA
2023/10/06 10:24:31 x509: cannot parse IP address of length 0
exit status 1
make: *** [Makefile:57: generate-certs] Error 1
```

ZMQ is broker-less, so it shouldn't need a service spun up in the docker-compose, unless I misunderstand the purpose (is it for spinning up mock brokers for the test?)

Reply:
I see. I tried running this branch locally and got this error. 
```sh
app-app-1      | /fleet-telemetry: error while loading shared libraries: libzmq.so.5: cannot open shared object file: No such file or directory
app-app-1 exited with code 127
```
does this help?

Reply:
This is helpful, I'll keep digging. Thanks!

Reply:
for `make generate-certs`, can you rebase your PR against latest main commit and see if the issue still persists?

Reply:
That fixed it, thanks!

Reply:
I'm having some issues with DNS inside the docker compose, any tips on how to fix it?

Reply:
can you elaborate on the issue?

Reply:
Seems like the app container isn't able to create sockets, I'm getting this error:

```
• [FAILED] [0.003 seconds]
Test messages [BeforeAll]
/go/src/fleet-telemetry/test/integration/integration_test.go:60
  reads vehicle data from consumer
  /go/src/fleet-telemetry/test/integration/integration_test.go:96

  Unexpected error:
      <*net.OpError | 0xc000124000>: {
          Op: "dial",
          Net: "tcp",
          Source: nil,
          Addr: nil,
          Err: <*net.DNSError | 0xc0001e6280>{
              Err: "server misbehaving",
              Name: "app",
              Server: "127.0.0.11:53",
              IsTimeout: false,
              IsTemporary: true,
              IsNotFound: false,
          },
      }
      dial tcp: lookup app on 127.0.0.11:53: server misbehaving
  occurred
  In [BeforeAll] at: /go/src/fleet-telemetry/test/integration/server_test.go:77
```

Some others have had issues with docker compose and dns as well, but I haven't found any solutions that work yet. Was wondering if anyone on your team has seen similar issues

Reply:
Seems to be due to changes on my branch, will continue looking into it (integration tests pass on main)

Reply:
it might be a combo of 
https://github.com/teslamotors/fleet-telemetry/pull/66#discussion_r1349187121 
and 
https://github.com/teslamotors/fleet-telemetry/pull/66#discussion_r1349134967
Unfortunately, I do not have more pointers than that

Reply:
some more observations. 
Expose zmq port here for integration test to pick up?
https://github.com/teslamotors/fleet-telemetry/blob/main/docker-compose.yml#L18
and then 
https://github.com/teslamotors/fleet-telemetry/pull/66/files#r1349377685